### PR TITLE
Changed translator logic to emit code on the fly, rather than manipulate AST.

### PR DESCRIPTION
**Pull request by azreika**
Previously, the translator cloned or otherwise manipulated AST objects to produce RAM code. This leads to analysis invalidation and hacky code to simulate desired behaviour. The main purpose of this PR is to eliminate any AST-lowering behaviour at the translator stage. Instead, the translator now exploits AST analyses to produce RAM code directly when needed.

Although this PR completes most of the work, cleanup is still required, particularly for provenance, for maintainability and to ensure all cases are covered. In particular, ADTs and aggregator support for provenance. This will be done in a future PR.

Nevertheless, the elimination of the bulk of AST-lowering mechanisms in the AST has major consequences on the translator codebase:
- ADTs are now handled as actual language features, translated directly to RAM
- The typestate introduced in PR #1725 is eliminated entirely from all AST objects - namely numeric constants, functors, binary constraints, and aggregators.
- AST cloning no longer occurs during translation
- `ast::ProvenanceNegation` and `ast::SubroutineArgument` are no longer needed, and have been purged

Other changes in this PR:
- The provenance clause translator now translates clauses when provenance is set, rather than subproof generation
- Subproof generation now handled by a new clause translator subclass - "ProvenanceSubproofGenerator" - significantly improving clarity
- Some other clean-ups

Next steps should introduce an abstract translator factory class, work towards finalising the structure of the translator, and clean up lingering hacks.

### CODE REVIEW

**Review by b-scholz — DISMISSED** [review]

File: src/ast/utility/Utils.cpp [hunk 1]
```diff
@@ -263,6 +263,24 @@ Clause* cloneHead(const Clause* clause) {
     return clone;
 }
 
+std::vector<ast::Atom*> reorderAtoms(
+        const std::vector<ast::Atom*>& atoms, const std::vector<unsigned int>& newOrder) {
```
Comment (mmcgr) [inline src/ast/utility/Utils.cpp:267]: This is already in the ast namespace

File: src/ast/utility/Utils.cpp [hunk 2]
```diff
@@ -263,6 +263,24 @@ Clause* cloneHead(const Clause* clause) {
...
+    // Validate given order
+    assert(newOrder.size() == atoms.size());
+    std::vector<unsigned int> nopOrder;
+    for (unsigned int i = 0; i < atoms.size(); i++) {
+        nopOrder.push_back(i);
+    }
+    assert(std::is_permutation(nopOrder.begin(), nopOrder.end(), newOrder.begin()));
+
+    // Create the result
+    std::vector<ast::Atom*> result(atoms.size());
```
Comment (mmcgr) [inline src/ast/utility/Utils.cpp:277]: This is already in the ast namespace

File: src/ast/utility/Utils.h [hunk 3]
```diff
@@ -242,6 +242,9 @@ bool isDeltaRelation(const QualifiedName& name);
  */
 Clause* cloneHead(const Clause* clause);
 
+std::vector<ast::Atom*> reorderAtoms(
+        const std::vector<ast::Atom*>& atoms, const std::vector<unsigned int>& newOrder);
```
Comment (mmcgr) [inline src/ast/utility/Utils.h:246]: This is already in the ast namespace

File: src/ast2ram/AstToRamTranslator.cpp [hunk 4]
```diff
@@ -678,26 +495,12 @@ Own<ram::Sequence> AstToRamTranslator::generateProgram(const ast::TranslationUni
...
-    removeADTs(tu);
-}
-
 Own<ram::TranslationUnit> AstToRamTranslator::translateUnit(ast::TranslationUnit& tu) {
-    // Start timer
-    auto ram_start = std::chrono::high_resolution_clock::now();
-
     /* -- Set-up -- */
-    // Set up the translator
+    auto ram_start = std::chrono::high_resolution_clock::now();
```
Comment (mmcgr) [inline src/ast2ram/AstToRamTranslator.cpp:500]: Is this used?

File: src/ast2ram/ClauseTranslator.h [hunk 5]
```diff
@@ -47,38 +48,50 @@ class Statement;
...
-    /** Operation levelling */
-    int addGeneratorLevel(const ast::Argument* arg);
-    int addOperatorLevel(const ast::Node* node);
+    Own<ValueIndex> valueIndex;
+
+    Own<ram::Statement> translateClause(const ast::Clause& clause);
+
+    /** Main clause translation */
+    virtual Own<ram::Statement> createRamFactQuery(const ast::Clause& clause) const;
+    virtual Own<ram::Statement> createRamRuleQuery(const ast::Clause& clause);
```
Comment (mmcgr) [inline src/ast2ram/ClauseTranslator.h:87]: Is the variations with virtual and const in the new methods intentional?

File: src/ast2ram/ProvenanceSubproofGenerator.cpp [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            continue;
+        }
+
+        if (auto condition = ConstraintTranslator::translate(context, symbolTable, *valueIndex, lit)) {
+            op = mk<ram::Filter>(std::move(condition), std::move(op));
+        }
+    }
+
+    for (const auto* con : constraints) {
+        if (auto condition = ConstraintTranslator::translate(context, symbolTable, *valueIndex, con)) {
```
Comment (mmcgr) [inline src/ast2ram/ProvenanceSubproofGenerator.cpp:110]: con -> constraint

File: src/ast2ram/ProvenanceSubproofGenerator.cpp [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+            values.push_back(ValueTranslator::translate(context, symbolTable, *valueIndex, var));
+            values.push_back(mk<ram::SubroutineArgument>(i));
+        } else if (const auto* func = dynamic_cast<const ast::Functor*>(arg)) {
+            values.push_back(ValueTranslator::translate(context, symbolTable, *valueIndex, func));
+            values.push_back(mk<ram::SubroutineArgument>(i));
+        } else if (const auto* rec = dynamic_cast<const ast::RecordInit*>(arg)) {
+            values.push_back(ValueTranslator::translate(context, symbolTable, *valueIndex, rec));
+            values.push_back(mk<ram::SubroutineArgument>(i));
+        } else if (const auto* adt = dynamic_cast<const ast::BranchInit*>(arg)) {
+            // TODO: fill this out like record arguments
```
Comment (mmcgr) [inline src/ast2ram/ProvenanceSubproofGenerator.cpp:229]: Whose TODO?

File: src/ast2ram/ProvenanceSubproofGenerator.cpp [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            auto lhs = ValueTranslator::translate(context, symbolTable, *valueIndex, func);
+            auto constraint = mk<ram::Constraint>(opEq, std::move(lhs), mk<ram::SubroutineArgument>(i));
+            op = mk<ram::Filter>(std::move(constraint), std::move(op));
+        } else if (const auto* rec = dynamic_cast<const ast::RecordInit*>(arg)) {
+            auto lhs = ValueTranslator::translate(context, symbolTable, *valueIndex, rec);
+            auto constraint = mk<ram::Constraint>(
+                    BinaryConstraintOp::EQ, std::move(lhs), mk<ram::SubroutineArgument>(i));
+            op = mk<ram::Filter>(std::move(constraint), std::move(op));
+        } else if (const auto* adt = dynamic_cast<const ast::BranchInit*>(arg)) {
+            // TODO: fill this out like record arguments
```
Comment (mmcgr) [inline src/ast2ram/ProvenanceSubproofGenerator.cpp:140]: Whose TODO?

**Comment by azreika** [comment]
Hey Jenkins, test this please x

**Comment by b-scholz** [comment]
More changes will follow after this PR. This is only to get the states out of the AST etc.

**Comment by b-scholz** [comment]
The build timed out.

**Comment by mmcgr** [comment]
The nullary !empty checks have been removed, so 'recursively' inserting into a nullary never terminates.

File: src/ast2ram/AstToRamTranslator.cpp (hunk 4 above)
Comment (azreika) [reply src/ast2ram/AstToRamTranslator.cpp:500]: Yep, in the debug report section later on.

File: src/ast2ram/ClauseTranslator.h (hunk 5 above)
Comment (azreika) [reply src/ast2ram/ClauseTranslator.h:87]: Yep. `createRamFactQuery` doesn't need to index the clause, while `createRamRuleQuery` does. Indexing is a state-changing function. If it becomes more natural (and there's time), I might just pass around the index as needed.

**Comment by azreika** [comment]
Cleaning up/looking into the final issues.

**Comment by azreika** [comment]
`atom-frequency` field seems to be missing in certain cases from profile output. Fix coming up.

**Review by b-scholz — APPROVED** [review]

**Pull request merged**